rhp/v4/server.go

```diff
@@ -669,29 +669,193 @@ func (s *Server) handleRPCRenewContract(stream net.Conn) error {
+	}
+	defer unlock()
+
+	// validate challenge signature
+	existing := state.Revision
+	if !req.ValidChallengeSignature(existing) {
+		return errorBadRequest("invalid challenge signature")
+	}
 
 	// validate the request
-	if err := req.Validate(s.hostKey.PublicKey(), tip, settings.MaxCollateral, settings.MaxContractDuration); err != nil {
+	settings := s.settings.RHP4Settings()
+	if err := req.Validate(s.hostKey.PublicKey(), state.Revision.ExpirationHeight, settings.MaxCollateral); err != nil {
 		return rhp4.NewRPCError(rhp4.ErrorCodeBadRequest, err.Error())
 	}
 
+	cs := s.chain.TipState()
+	renewal := rhp4.RefreshContract(existing, prices, req.Refresh)
+	renterCost, hostCost := rhp4.RefreshCost(cs, prices, renewal, req.MinerFee)
+	renewalTxn := types.V2Transaction{
+		MinerFee: req.MinerFee,
+	}
+
+	// add the renter inputs
+	var renterInputSum types.Currency
+	for _, si := range req.RenterInputs {
+		renewalTxn.SiacoinInputs = append(renewalTxn.SiacoinInputs, types.V2SiacoinInput{
+			Parent: si,
+		})
+		renterInputSum = renterInputSum.Add(si.SiacoinOutput.Value)
+	}
+
+	if n := renterInputSum.Cmp(renterCost); n < 0 {
+		return errorBadRequest("expected renter to fund %v, got %v", renterInputSum, renterCost)
+	} else if n > 0 {
+		// if the renter added too much, add a change output
+		renewalTxn.SiacoinOutputs = append(renewalTxn.SiacoinOutputs, types.SiacoinOutput{
+			Address: renewal.NewContract.RenterOutput.Address,
+			Value:   renterInputSum.Sub(renterCost),
+		})
+	}
+
+	elementBasis, fce, err := s.contractor.ContractElement(req.Refresh.ContractID)
+	if err != nil {
+		return fmt.Errorf("failed to get contract element: %w", err)
+	}
+
+	basis, toSign, err := s.wallet.FundV2Transaction(&renewalTxn, hostCost, true)
+	if errors.Is(err, wallet.ErrNotEnoughFunds) {
+		return rhp4.ErrHostFundError
+	} else if err != nil {
+		return fmt.Errorf("failed to fund transaction: %w", err)
+	}
+
+	// update renter inputs to reflect our chain state
+	if basis != req.Basis {
+		hostInputs := renewalTxn.SiacoinInputs[len(renewalTxn.SiacoinInputs)-len(req.RenterInputs):]
+		renewalTxn.SiacoinInputs = renewalTxn.SiacoinInputs[:len(renewalTxn.SiacoinInputs)-len(req.RenterInputs)]
+		updated, err := s.chain.UpdateV2TransactionSet([]types.V2Transaction{renewalTxn}, req.Basis, basis)
+		if err != nil {
+			return errorBadRequest("failed to update renter inputs from %q to %q: %v", req.Basis, basis, err)
+		}
+		renewalTxn = updated[0]
+		renewalTxn.SiacoinInputs = append(renewalTxn.SiacoinInputs, hostInputs...)
+	}
+
+	if elementBasis != basis {
+		tempTxn := types.V2Transaction{
+			FileContractResolutions: []types.V2FileContractResolution{
+				{Parent: fce, Resolution: &renewal},
+			},
+		}
+		updated, err := s.chain.UpdateV2TransactionSet([]types.V2Transaction{tempTxn}, elementBasis, basis)
+		if err != nil {
+			return fmt.Errorf("failed to update contract element: %w", err)
```

Comment:
If we return a bad request when we update the renter inputs to reflect the chain state, shouldn't we return one here as well?

Reply:
Bad request should only be for issues with parameters from the renter. This error should be a host error since all of the input parameters are from the host.